### DISCUSSION TITLE
GenCExpo Mk.2

### DISCUSSION THREAD

**Discussion by dogfuntom** [dogfuntom/GenCExpo#24] [Ideas]
## The problem with Mk.1

The approach was Unity first. All other things being equal, it would be the easiest approach. But, in case of gen-expo, things are not equal.

Heavy interaction with JS side is unavoidable, entailing not only JS side existence but also existence of a big chunk of logic in between them.

However, no code can truly be in between, it should be JS or C#. This choice is not just about JS being more native to the end platform. It's also about what side is the leader and what has a passive role.

It turns out that the JS is the leader: it stores, initializes, renders, frame-captures, pauses and restarts the works. Unity/C# side is just a fancy alternative to a list or a grid of works.

## The new approach

The new approach should be JS first. However, first, it's easier to list what responsibilities are left to Unity side rather than list the JS side's responsibilities:

1. Pull the works and the configuration (JSON) from the JS side.
2. Show the works on the screen.
3. Push necessary data to JS side (data should flow in this direction only when there's no alternative to this).

Consequently, JS side should:

- Load the works dynamically. It should be possible to add a work without even having Unity editor installed.
- Provide the API for C# side to use, but as minimal as possible. If JS side can handle something itself, then so it should do.
- Have a mode of running without initializing Unity player, useful for debugging and testing its own internal code.
- Have a mode of running from dynamic configuration (instead of statically served one). With this, editing configuration (like adding a work) should be quick while avoiding commit spam in git.

# TL;DR

If C# side needs to have access to A, B, C to do job X, then JS side must provide A, and B, and C as an API. It's a craft. Instead, it should just do the job X itself, if it can. And usually it can.

Additionally, it's the only way to allow iterating on configuration (like when adding a work) quickly, without compilation or multiple trial-and-error commits.
